### PR TITLE
Upgrades do_after and progress bars.

### LINKED ISSUE #14049: Mining with pick in off-hand?
Don't know when it changed, or if it's even intentional, but it seems like the way that mining is done has been fundamentally changed within the past couple of days.

Normal behavior: Mining with a pick in your offhand means you can just push up against a direction and start automatically mining, rather than having to click the asteroid to mine

Current behavior: Mining with a pick in your offhand will show the progress bar for about a half second and then abort out with no message.

### LINKED ISSUE #14052: Either do_after is broken or the players are idiots
[That's what went through my head when I read this](https://tgstation13.org/phpBB/viewtopic.php?f=10&t=5487#p142842)
`If the drain progress is about half-way complete, it doesn't matter if they drag you away because it'll still finish and the revenant gets the points. I've done it multiple times to people.`
For reference, revenant draining uses do_after

### PR DESCRIPTION
**Pull request by MrsTonedOne**
Fixes #14052

Fixes #14049

They will now always tick at 1ds, rather then by delay/numofticks.

Reworked my change with lastmove, SHOULD be fine now, as failed moves won't count.

do_mob and do_after moved to __helpers/mob.dm

Progress bar system made a datum and moved to datums/

Progress bars now have 20 icon states, or 5% resolution. 1 for each of the 20 pixels in the actual bar area.

Renamed progress bar icon file to........... progressbars

https://tgstation13.org/msoshit/progbar.flv

https://www.youtube.com/watch?v=ZaxuYun9ikA for you fucks who can't be assed to view a .flv file in a video player

:cl:
tweak: Progress bars now have 20 states instead of 5.
fix: Bump mining should be fixed.
tweak: Long actions now checks for movement/changed hands/etc more often to prevent edge cases.
tweak: Multiple progress bars will now stack properly, rather than fight each other for focus.
/:cl:

~~~I have a few more things to do, to fix bugs before this is ready.~~~ I'll do it in another PR, this is ready

### CODE REVIEW

**Comment by Iamgoofball** [comment]
>flv

**Comment by RemieRichards** [comment]
:+1:

File: code/__HELPERS/mobs.dm [hunk 1]
```diff
@@ -139,4 +139,90 @@ Proc for attack log creation, because really why not
...
+	. = 1
+	while (world.time < endtime)
+		if (progress)
+			progbar.update(endtime - world.time)
+		if(!user || !target)
+			. = 0
+			break
+		if(uninterruptible)
+			continue
+		if(user.loc != user_loc || user.last_move || target.loc != target_loc || target.last_move != null || user.get_active_hand() != holding || user.incapacitated() || user.lying )
```
Comment (phil235) [inline code/__HELPERS/mobs.dm]: why not use the same method as with do_after() and have `!target.last_move` instead of `target.last_move != null` ?

Comment (MrStonedOne) [inline code/__HELPERS/mobs.dm]: must have missed it

**Pull request merged**